File: lib/teien/actor.rb
require_relative './physics'

module Teien
  class Actor < Bullet::BtMotionState
    attr_accessor :name
    attr_accessor :state

    attr_accessor :physics_info
    attr_accessor :ext_info
    
    def initialize(name, physics_info = nil, ext_info = nil)
      super() # This is ultimate important for detecting overloaded method.
      
      @name = name
      @state = "none"
      @physics_info = check_physics_info(physics_info)
      @ext_info = ext_info ? ext_info : {}

      @transform = Bullet::BtTransform.new()
      @transform.set_identity()

      @rigid_body = nil
    end

    def check_physics_info(physics_info)
      @physics_info = {}
      @physics_info[:use_physics] = physics_info == nil ? false : true
      return @physics_info if (physics_info == nil)
      
      @physics_info[:mass] = physics_info[:mass] == nil ? 0 : physics_info[:mass]
      @physics_info[:restitution] = physics_info[:restitution] == nil ? 0.2 : physics_info[:restitution]
      @physics_info[:friction] = physics_info[:friction] == nil ? 1.0 : physics_info[:friction]
      @physics_info[:linear_damping] = physics_info[:linear_damping] == nil ? 0.0 : physics_info[:linear_damping]
      @physics_info[:angular_damping] = physics_info[:angular_damping] == nil ? 0.0 : physics_info[:angular_damping]
      
      @physics_info
    end

    def setup(physics)
      if(@rigid_body)
        @rigid_body.set_restitution(@physics_info[:restitution])
        @rigid_body.set_friction(@physics_info[:friction])
        @rigid_body.set_damping(@physics_info[:linear_damping],
                                @physics_info[:angular_damping])
=begin         
        rigid_body.setAngularFactor(new Ammo.btVector3(actorInfo.angularFactor.x,
                                                      actorInfo.angularFactor.y,
                                                      actorInfo.angularFactor.z));
=end
      end
    end
    
    def set_world_transform(worldTrans)
      @transform = Bullet::BtTransform.new(worldTrans)
#      rot = @transform.get_rotation();
#      puts "set_world_transform::rot(#{rot.x}, #{rot.y}, #{rot.z}, #{rot.w})"
    end

    def get_world_transform(worldTrans)
#      @transform = @rigid_body.get_center_of_mass_transform() if (@rigid_body)
      worldTrans.set_origin(@transform.get_origin())
      worldTrans.set_rotation(@transform.get_rotation())
    end
    
    def set_position(vec)
      @transform.set_origin(Bullet::BtVector3.new(vec.x, vec.y, vec.z))
      @rigid_body.set_center_of_mass_transform(@transform) if @rigid_body
    end

    def set_rotation(quat)
      @transform.set_rotation(Bullet::BtQuaternion.new(quat.x, quat.y, quat.z, quat.w))
      @rigid_body.set_center_of_mass_transform(@transform) if @rigid_body
    end

    def set_linear_velocity(vec)
      @rigid_body.set_linearr_velocity(vec) if @rigid_body
    end

    def get_linear_velocity()
      return @rigid_body.get_linear_velocity() if @rigid_body
      return Bullet::BtVector3.new
    end    

    def get_angular_velocity()
      return @rigid_body.get_angular_velocity() if @rigid_body
      return Bullet::BtVector3.new
    end    

    def to_hash
      pos = @transform.get_origin()
      rot = @transform.get_rotation()
      linear_vel = get_linear_velocity()
      angular_vel = get_angular_velocity() 
      
      { 
        :name => @name,
        :state => @state,
        :physics_info => @physics_info,
        :transform => {
          :position => {
            :x => pos.x,
            :y => pos.y,
            :z => pos.z
          },
          :rotation => {
            :x => rot.x,
            :y => rot.y,
            :z => rot.z,
            :w => rot.w
          }
        },
        :linear_vel => {
          :x => linear_vel.x,
          :y => linear_vel.y,
          :z => linear_vel.z          
        },
        :angular_vel => {
          :x => angular_vel.x,
          :y => angular_vel.y,
          :z => angular_vel.z          
        },
        :ext_info => @ext_info
      }
    end

    def from_hash(hash)
      @name = hash[:name]
      @state = hash[:state]
      @transform.from_hash(hash[:transform])
      @ext_info = hash[:ext_info]
    end

    def update(delta)
    end
  end
end

File: public/teien/actor.js
teien.Actor = function(param) {
    this.param = param;
    
	this.name = param.name;
    this.state = param.state;
    this.physics_info = param.physics_info;    
    this.ext_info = param.ext_info;

    this.transform = new Ammo.btTransform();
    this.transform.setIdentity();
    
    this.rigidBody = null;
    
    this.observers = [];
};

teien.Actor.prototype.setup = function(physics) {
    if (this.rigidBody){
        this.rigidBody.setRestitution(this.physics_info.restitution);
        this.rigidBody.setFriction(this.physics_info.friction);
        this.rigidBody.setDamping(this.physics_info.linear_damping,
                                  this.physics_info.angular_damping);
    }
    this.fromHash(this.param);
};

// vec: teien.Vector3D
teien.Actor.prototype.setPosition = function(vec) {
    this.transform.setOrigin(new Ammo.btVector3(vec.x, vec.y, vec.z));
    if (this.rigidBody) {
        this.rigidBody.setCenterOfMassTransform(this.transform);
	    //this.rigidBody.setWorldTransform(this.transform);        
        this.rigidBody.activate();
    }
};

teien.Actor.prototype.setInterpolatePosition = function(vec) {
    var oldVec = this.getPosition();
    var ipVec = new teien.Vector3D((vec.x + oldVec.x) / 2, 
                                   (vec.y + oldVec.y) / 2, 
                                   (vec.z + oldVec.z) / 2);
    //console.log("(" + ipVec.x() + ", " + ipVec.y() + ", " + ipVec.z() +")");

    this.setPosition(ipVec);
};


// quat: teien.Quaternion
teien.Actor.prototype.setRotation = function(quat) {
    this.transform.setRotation(new Ammo.btQuaternion(quat.x, quat.y, quat.z, quat.w));
    if (this.rigidBody) {    
        this.rigidBody.setCenterOfMassTransform(this.transform);
	    //this.rigidBody.setWorldTransform(this.transform);                
        this.rigidBody.activate();
    }
};

teien.Actor.prototype.setLinearVelocity = function(vec) {
    if (this.rigidBody){
        this.rigidBody.setLinearVelocity(new Ammo.btVector3(vec.x, vec.y, vec.z));
        this.rigidBody.activate();
    }
};

teien.Actor.prototype.setAngularVelocity = function(vec) {
    if (this.rigidBody){
        this.rigidBody.setAngularVelocity(new Ammo.btVector3(vec.x, vec.y, vec.z));
        this.rigidBody.activate();
    }
};

teien.Actor.prototype.getPosition = function() {
    var pos = this.transform.getOrigin();
	return teien.Vector3D.createFromAmmo(pos);
};

teien.Actor.prototype.getRotation = function() {
	return teien.Quaternion.createFromAmmo(this.transform.getRotation())
};

teien.Actor.prototype.getLinearVelocity = function() {
    return this.rigidBody.getLinearVelocity();
};

/* These two functions are not supported by Ammo.
teien.Actor.prototype.setWorldTransform = function(worldTrans){
    console.log("setWorldTransform is called.\n");
    this.transform = new Ammo.btTransform(worldTrans);
};

teien.Actor.prototype.getWorldTransform = function(worldTrans){
    console.log("getWorldTransform is called.\n");
};
*/

teien.Actor.prototype.fromHash = function(param){
	this.name = param.name;
    this.state = param.state;
    this.physics_info = param.physics_info;
    this.ext_info = param.ext_info;
    
    var pos = new teien.Vector3D(param.transform.position.x,
                                 param.transform.position.y,
                                 param.transform.position.z);
    this.setPosition(pos);
    //this.setInterpolatePosition(pos);    
    var rot = new teien.Quaternion(param.transform.rotation.x,
                                   param.transform.rotation.y,
                                   param.transform.rotation.z,
                                   param.transform.rotation.w);
    this.setRotation(rot);

    this.setLinearVelocity(new teien.Vector3D(param.linear_vel.x,
                                              param.linear_vel.y,
                                              param.linear_vel.z));
    
    this.setAngularVelocity(new teien.Vector3D(param.angular_vel.x,
                                               param.angular_vel.y,
                                               param.angular_vel.z));
    pos = this.getPosition();
};

teien.Actor.prototype.toHash = function(){
    var hash = {};
	hash.name = this.name;
    hash.state = this.state;
    hash.physics_info = this.physics_info;
    hash.ext_info = this.ext_info;

    var trans = {};
    var pos = this.transform.getOrigin();
    trans.position = {
        x: pos.x(),
        y: pos.y(),
        z: pos.z()
    };
    var rot = this.transform.getRotation();    
    trans.rotation = {
        x: rot.x(),
        y: rot.y(),
        z: rot.z(),
        w: rot.w()
    };
    hash.transform = trans;

    if (this.rigidBody) {
        var l_v = this.rigidBody.getLinearVelocity();
        hash.linear_vel = {
            x: l_v.x(),
            y: l_v.y(),
            z: l_v.z()
        };
    }

    if (this.rigidBody) {    
        var a_v = this.rigidBody.getAngularVelocity();
        hash.angular_vel = {
            x: a_v.x(),
            y: a_v.y(),
            z: a_v.z()
        };
    }

    return hash;
};


/*
teien.Actor.prototype.applyImpulse = function(imp, rel) {
	if (rel === undefined)
	    rel = new exports.Vector3D(0, 0, 0);
    
	this.physicsState.applyImpulse(imp, rel);
};
*/

teien.Actor.prototype.updateTransform = function(delta) {

    if (this.rigidBody){
        this.transform = this.rigidBody.getCenterOfMassTransform();
        // This function don't return the right value of @mass == 0.
        //this.rigidBody.getMotionState().getWorldTransform(this.transform); 
        //console.log(this.transform.getOrigin().y() + "\n");
    }
    //this.notify(delta);    
};

/*
// Objserver
teien.Actor.prototype.attach = function(observer){
    for(i=0; i < this.observers.length; i++){
        if(this.observers[i] == observer)
            return;
    }
    this.observers.push(observer);
};

teien.Actor.prototype.detach = function(observer){
    for(i=0; i < this.observers.length; i++){
        if(this.observers[i] == observer){
            this.observers.splice(i--, 1);
        }
    }    
};

teien.Actor.prototype.notify = function(delta){
    for(i=0; i < this.observers.length; i++){
        this.observers[i].update(delta);
    }
}
*/
